### DISCUSSION TITLE
Trying to implement pagination with Prisma/Nexus. How can I return the total count of posts in my blog?

### DISCUSSION THREAD

**Discussion by lumenwrites** [prisma/prisma#8103] [Q&A] [answered]
Hi! I have a `posts` query that returns the list of posts filtered by author/tag/search/etc:

```ts
context.prisma.post.findMany({
          where: {
            authorId: authorId,
            published: args.published || undefined,
            ...searchAndFilter,
          },
          orderBy: [{ score: 'desc' }],
          take: args.take,
          skip: args.skip,
        })
```

In order to implement pagination, I also need to know the total number of posts my query returns after filtering (so that I can know the number of pages to display).

I'm finding this surprisingly difficult. What is the right way to return the total number of posts along with the list of posts?

**Comment by ryands17** [comment, marked as answer]
@lumenwrites 👋 

You need two queries for this:

```ts
const [posts, postCount] = await prisma.$transaction([
    context.prisma.post.findMany({
          where: {
            authorId: authorId,
            published: args.published || undefined,
            ...search,
            ...tagFilter,
            ...topicFilter,
          },
          orderBy: [{ score: 'desc' }],
          take: args.take || undefined,
          skip: args.skip || undefined,
      }),
      context.prisma.post.count()
])
```

This should give you the desired output and you can return this from your resolver.

**Reply by lumenwrites** [reply]
Thank you very much! Two questions:

1. How do I query this? Just running the query the way I used to, kinda like this, doesn't seem to work:

```graphql
query {
  posts (
    searchString: "post",
  ) {
    title
    tags {
      name
      slug
    }
  }
}
```

I used to return `findMany` from the resolver:


```ts
export const PostQueries = extendType({
  type: 'Query',
  definition(t) {
    t.nonNull.list.nonNull.field('posts', {
      type: 'Post',
      args: {
        [...]
      },
      resolve: async (_parent, args, context: Context) => {
        [...]
        return context.prisma.post.findMany({
          where: {
            [...]
          },
          orderBy: [{ score: 'desc' }],
          take: args.take,
          skip: args.skip,
          
        })
      },
    })
```

but do I now need to await? Do I need a separate nexus return type for this? Not just `Post`, but, somehow, `Post` with an extra field for the count?

2. In your code, the `.count()` would return the total count of posts, unfiltered, right? If I want to count the posts that were found using the search (or filtering by author or by tags), do I just need t pass the same exact arguments to `.count()` as I did to `.findMany()`? (like `where:{...}`, `take`, `skip`, etc.)

**Reply by ryands17** [reply]
1. Your return type would be different. You need to create a new type that has Posts along with the count of them.
2. Yes it would and you can pass the same arguments. I assume the arguments for count are different than the arguments of the main query to find Posts because if the filters are same then it doesn't make sense to count them.

**Reply by lumenwrites** [reply]
Thank you, that makes sense! Yes, I'm just realizing that I don't need `take` and `skip` in the count, I just need to pass the filters to where.

One last thing, I'm trying to create a custom nexus type for the return, it looks like this:

```
export const PostsWithCount = objectType({
  name: 'PostsWithCount',
  definition(t) {
    t.field('filteredPosts', { type: 'Post' })
    t.int('count')
  },
})
```

I think I'm doing it wrong, because the result of this query:

```
query {
  posts {
    filteredPosts {
      title
    }
    count
  }
}
```

is:

```
{
  "data": {
    "posts": [
      {
        "filteredPosts": null,
        "count": null
      },
      {
        "filteredPosts": null,
        "count": null
      }
    ]
  }
}
```

So it doesn't throw an error, but doesn't work correctly either. Could you help me figure out what am I doing wrong?

When I'm doing a printing out the results of the transaction, I see the list of posts and their count displayed correctly in the terminal, so I think my resolver does return the correct results:

```
        const [posts, postCount] = await context.prisma.$transaction([...])
        console.log([posts, postCount])
        return [posts, postCount]
```

**Reply by ryands17** [reply]
The type should be this:

```ts
export const PostsWithCount = objectType({
  name: 'PostsWithCount',
  definition(t) {
    t.list.field('posts', { type: 'Post' })
    t.int('count')
  },
})
```

And the return type of the query should simply be: `PostsWithCount`. Your resolver should also return an object, not an array:

```ts
const [posts, count] = await context.prisma.$transaction([...])

return { posts, count }
```

**Reply by lumenwrites** [reply]
Hmm, when I return an object instead of the array, I'm getting:

>"Expected Iterable, but did not find one for field \"Query.posts\"."

**Reply by ryands17** [reply]
Could you share the entire Query field?

**Reply by lumenwrites** [reply]
The query is [here](https://github.com/lumenwrites/sketchclub/blob/master/backend/nexus/postQueries.ts) (called `posts`).

The type is [here](https://github.com/lumenwrites/sketchclub/blob/master/backend/nexus/postTypes.ts) (`PostsWithCount`).

**Reply by ryands17** [reply]
```ts
t.nonNull.field('posts', {
      type: 'PostsWithCount',
      args: {
        username: stringArg(),
        published: booleanArg(),
        searchString: stringArg(),
        tagSlug: stringArg(),
        topicSlug: stringArg(),
        skip: intArg(),
        take: intArg(),
      },
      resolve: async (_parent, args, context: Context) => {
        console.log('Get posts', args)
        // Filter posts by user
        let authorId
        if (args.username) {
          const author = await context.prisma.user.findUnique({
            where: { username: args.username },
          })
          authorId = author.id
        }
        // Filter by tag
        const tagFilter = args.tagSlug ? {
          tags: { some: { slug: args.tagSlug } }
        } : {}
        // Filter by topic
        const topicFilter = args.topicSlug ? {
          topic: { slug: { contains: args.topicSlug } }
        } : {}
        // Search through posts
        const search = args.searchString ? {
          OR: [
            { title: { contains: args.searchString, mode: "insensitive", } },
            { body: { contains: args.searchString, mode: "insensitive", } },
            { tags: { some: { name: { contains: args.searchString, mode: "insensitive", } } } },
            { topic: { name: { contains: args.searchString, mode: "insensitive", } } }
            // TODO: search in username too
          ],
        } : {}
        // Sort by hot
        // https://medium.com/hacking-and-gonzo/how-hacker-news-ranking-algorithm-works-1d9b0cf2c08d
        // https://medium.com/hacking-and-gonzo/how-reddit-ranking-algorithms-work-ef111e33d0d9
        // const age = new Date() - date
        // const gravity = 1.8
        // const rankScore = (score - 1) / (age + 2) ** gravity
        // console.log('Posts resolver', context.prisma.post.findMany())
        const allFilters = {
          authorId: authorId,
          published: args.published || undefined,
          ...search,
          ...tagFilter,
          ...topicFilter,
        }
        const [posts, postCount] = await context.prisma.$transaction([
          context.prisma.post.findMany({
            where: allFilters,
            orderBy: [{ score: 'desc' }],
            take: args.take || undefined,
            skip: args.skip || undefined,
          }),
          context.prisma.post.count({
            where: allFilters
          })
        ])
        console.log([posts, postCount])
        return { posts, postCount }
      },
    })
```

The list isn't required as it's a single object in itself.

**Reply by lumenwrites** [reply]
Thank you!

The query now runs, but still returns the nulls:

```
{
  "data": {
    "getPosts": {
      "posts": {
        "body": null
      },
      "count": null
    }
  }
}
```

(I have renamed the query from `posts` to `getPosts` for clarity.)

the `{ posts, postCount }` are definitely not empty, conosle.log still prints out a list of posts and the total number of posts as I would expect...

**Reply by ryands17** [reply]
Could you check if the log of `posts` and `count` matches the object in Nexus?

**Reply by lumenwrites** [reply]
Oh, wait, sorry, figured it out and fixed it!

The return type should look like this:

```
export const PostsWithCount = objectType({
  name: 'PostsWithCount',
  definition(t) {
    t.list.field('posts', { type: 'Post' })
    t.int('postCount')
  },
})
```

I was experimenting and deleted `list` from `t.list.field('posts', { type: 'Post' })`, and ` t.int('postCount')` used to be ` t.int('count')`. Silly mistake, now everything works.

Thank you so much! This is awesome! I'm very happy =)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
